rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+- Start Date: 2019-06-29
+- RFC PR: (leave this empty)
+- Kibana Issue: https://github.com/elastic/kibana/issues/33779
+
+# Summary
+
+Http Service in New platform should provide the ability to execute some logic in response to an incoming request and send the result of this operation back.
+
+# Basic example
+Declaring a route handler for `/url` endpoint:
+```typescript
+router.get(
+  { path: '/url', ...[otherRouteParameters] },
+  (context: Context, request: KibanaRequest, t: KibanaResponseToolkit) => {
+   // logic to handle request ...
+   return t.ok(result);
+);
+
+```
+
+# Motivation
+The new platform is built with library-agnostic philosophy and we cannot transfer the current solution for Network layer from Hapi. To avoid vendor lock-in in the future, we have to define route handler logic and request/response objects formats that can be implemented in any low-level library such as Express, Hapi, etc. It means that we are going to operate our own abstractions for such Http domain entities as Router, Route, Route Handler, Request, Response.
+
+# Detailed design
+The new platform doesn't support the Legacy platform `Route Handler` format nor exposes implementation details, such as [Hapi.ResponseToolkit](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/hapi/v17/index.d.ts#L984).
+Rather `Route Handler` in New platform has the next signature:
+```typescript
+type RequestHandler = (
+  context: Context,
+  request: KibanaRequest,
+  t: KibanaResponseToolkit
+) => KibanaResponse | Promise<KibanaResponse>;
```

Comment:
Given that HTTP requests are inherently async, I'm not sure if there is any benefit of a synchronouse response. Maybe just one return type? Like so:

```ts
) => Promise<KibanaResponse>;
```

Reply:
A synchronous handler is a perfectly valid function. Isn't it?
```js
handler(context, request, t) => t.redirect('url/')
```